DOC: update doc-string in Index.rename

Moons08 opened the pull request:
updated doc-strings with examples 
- [x] closes #20787
- N/A tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- N/A whatsnew entry

Moons08 committed:
Update doc-string in .rename

jbrockmendel commented on pandas/core/indexes/base.py:
```diff
@@ -1340,17 +1340,39 @@ def set_names(self, names, level=None, inplace=False):
+        Int64Index([1, 2, 3, 4], dtype='int64', name='foo')
+        >>> idx.rename('bar')
+        Int64Index([1, 2, 3, 4], dtype='int64', name='bar')
+        >>> idx = MultiIndex.from_tuples([(1, u'one'), (1, u'two'),
```
Should this have a newline to denote a separate example?

jorisvandenbossche commented on pandas/core/indexes/base.py:
```diff
@@ -1340,17 +1345,40 @@ def set_names(self, names, level=None, inplace=False):
     def rename(self, name, inplace=False):
         """
         Set new names on index. Defaults to returning new index.
+        Length of names must match number of levels in MultiIndex.
```
Leave a blank line between the two lines above (the first line of the docstring should be a single line, see http://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html#docstring for details)

Moons08 replied:
I changed some line with your suggestion. 
and I put `\` in the MultiIndex example to avoid error. and what happened in validation error.. 
thank you so much.

```python
################################################################################
####################### Docstring (pandas.Index.rename)  #######################
################################################################################

Set new names on index. Defaults to returning new index.

Length of names must match number of levels in MultiIndex.

Parameters
----------
name : str or sequence
    name(s) to set
inplace : bool
    if True, mutates in place

Returns
-------
new index (of same type and class...etc) [if inplace, returns None]

Examples
--------
>>> idx = pd.Index([1, 2, 3, 4], name='foo')
>>> idx.rename('bar')
Int64Index([1, 2, 3, 4], dtype='int64', name='bar')

>>> idx = pd.MultiIndex.from_tuples([(1, u'one'), (1, u'two'),                                            (2, u'one'), (2, u'two')],                                            names=['foo', 'bar'])
>>> idx.rename(['bar', None], inplace=True)
>>> idx
MultiIndex(levels=[[1, 2], ['one', 'two']],
           labels=[[0, 0, 1, 1], [0, 1, 0, 1]],
           names=['bar', None])
>>> idx.rename(['bar'])
ValueError: Length of names must match number of levels in MultiIndex.

See also
--------
set_names : able to set new names partially and by level

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	Errors in parameters section
		Parameter "name" description should start with capital letter
		Parameter "name" description should finish with "."
		Parameter "inplace" description should start with capital letter
		Parameter "inplace" description should finish with "."
	Examples do not pass tests

################################################################################
################################### Doctests ###################################
################################################################################

**********************************************************************
Line 29, in pandas.Index.rename
Failed example:
    idx.rename(['bar'])
Exception raised:
    Traceback (most recent call last):
      File "/home/mk/anaconda3/lib/python3.6/doctest.py", line 1330, in __run
        compileflags, 1), test.globs)
      File "<doctest pandas.Index.rename[5]>", line 1, in <module>
        idx.rename(['bar'])
      File "/home/mk/documents/dev/pandas-mk/pandas/core/indexes/base.py", line 1341, in set_names
        idx._set_names(names, level=level)
      File "/home/mk/documents/dev/pandas-mk/pandas/core/indexes/multi.py", line 672, in _set_names
        raise ValueError('Length of names must match number of levels in '
    ValueError: Length of names must match number of levels in MultiIndex.
```

jorisvandenbossche replied:
Can you fix the "Errors found:" ? (capitalization and punctuation)

jorisvandenbossche commented on pandas/core/indexes/base.py:
```diff
@@ -1340,17 +1345,40 @@ def set_names(self, names, level=None, inplace=False):
+
+        Examples
+        --------
+        >>> idx = Index([1, 2, 3, 4], name='foo')
```
Can you use `pd.` for pandas objects? (but no need to import it)

jorisvandenbossche commented on pandas/core/indexes/base.py:
```diff
@@ -1340,17 +1345,40 @@ def set_names(self, names, level=None, inplace=False):
+        >>> idx = MultiIndex.from_tuples([(1, u'one'), (1, u'two'),
+                                          (2, u'one'), (2, u'two')],
+                                          names=['foo', 'bar'])
+        >>> idx.rename(['bar', None], inplace=True)
```
I would personally not show the inplace argument

jorisvandenbossche commented on pandas/core/indexes/base.py:
```diff
@@ -1340,17 +1345,40 @@ def set_names(self, names, level=None, inplace=False):
         ----------
-        name : str or list
-            name to set
+        name : str or sequence
```
is there a reason you changed 'list' to 'sequence' ? 
I know it technically doest accept any list, but in index names there *can* be a difference between a list (multiple names) and a tuple (single name), so I would rather keep it as list. 
But it is true that a name is not limited to a string, so maybe we should use the more general "label" (and then "label, or list of labels")

Moons08 replied:
The reason why I used ‘sequence’ is that rename function takes not only list but also tuple in multiple names. 
I am not sure it is intended, but it is.

jorisvandenbossche replied:
Yes, it is intended. But as I tried to explain, a tuple can also be a single name (so you can eg give a list of tuples). Therefore, I don't think the change is an improvement in clarity for the user. 
I would strictly use 'list' for referring to the ability to give multiple names.

Moons08 replied:
but it works like this.
```python
>>> idx = pd.MultiIndex(levels=[[1, 2], ['one', 'two']],\
...                     labels=[[0, 0, 1, 1], [0, 1, 0, 1]],\
...                     names=['bar', None])
>>> idx
MultiIndex(levels=[[1, 2], ['one', 'two']],
           labels=[[0, 0, 1, 1], [0, 1, 0, 1]],
           names=['bar', None])
>>> idx.rename(('foo', 'bar'))
MultiIndex(levels=[[1, 2], ['one', 'two']],
           labels=[[0, 0, 1, 1], [0, 1, 0, 1]],
           names=['foo', 'bar'])
```

jorisvandenbossche replied:
Yes, it know it works, but we should not encourage that. 
For a single Index, there is a difference:

```
In [7]: idx = pd.Index([1, 2, 3, 4], name='foo')

In [8]: idx.rename(('bar', 'test'))
Out[8]: Int64Index([1, 2, 3, 4], dtype='int64', name=('bar', 'test'))

In [9]: idx.rename(['bar', 'test'])
...
TypeError: Int64Index.name must be a hashable type
```

So I would clearly state in our docstring one should use a list for multiple names (and not a sequence in general), to not confuse between those.

Moons08 replied:
oh, I got it. so `list` means not only data type. thank you.

jorisvandenbossche reviewed (commented):
Thanks for the PR!

Can you check your docstring with `python scripts/validate_docstrings.py pandas.Index.rename` (and paste the output of it here in a comment) ?

jorisvandenbossche commented on pandas/core/indexes/base.py:
```diff
@@ -1340,17 +1345,40 @@ def set_names(self, names, level=None, inplace=False):
+
+        >>> idx = MultiIndex.from_tuples([(1, u'one'), (1, u'two'),
+                                          (2, u'one'), (2, u'two')],
+                                          names=['foo', 'bar'])
```
To get this passing for the doctests, you need to add `...` at the beginning of the line, as you see it in the console:

```
        >>> idx = MultiIndex.from_tuples([(1, u'one'), (1, u'two'),
        ...                               (2, u'one'), (2, u'two')],
        ...                               names=['foo', 'bar'])
```

jorisvandenbossche commented on pandas/core/indexes/base.py:
```diff
@@ -1340,17 +1345,40 @@ def set_names(self, names, level=None, inplace=False):
+                   labels=[[0, 0, 1, 1], [0, 1, 0, 1]],
+                   names=['bar', None])
+        >>> idx.rename(['bar'])
+        ValueError: Length of names must match number of levels in MultiIndex.
```
To get this passing the doctests, see http://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html#tips-for-getting-your-examples-pass-the-doctests

Moons08 commented:
I fixed some details in `Index.rename` and `Index.set_names`

```python
################################################################################
####################### Docstring (pandas.Index.rename)  #######################
################################################################################

Set new names on index. Defaults to returning new index.

Length of names must match number of levels in MultiIndex.

Parameters
----------
name : str or list
    Name(s) to set.
inplace : bool
    If True, mutates in place.

Returns
-------
new index (of same type and class...etc) [if inplace, returns None]

Examples
--------
>>> idx = pd.Index([1, 2, 3, 4], name='foo')
>>> idx.rename('bar')
Int64Index([1, 2, 3, 4], dtype='int64', name='bar')

>>> idx = pd.MultiIndex.from_tuples([(1, u'one'), (1, u'two'),
...                                  (2, u'one'), (2, u'two')],
...                                  names=['foo', 'bar'])
>>> idx.rename(['bar', None])
MultiIndex(levels=[[1, 2], ['one', 'two']],
           labels=[[0, 0, 1, 1], [0, 1, 0, 1]],
           names=['bar', None])
>>> idx.rename(['bar'])
Traceback (most recent call last):
ValueError: Length of names must match number of levels in MultiIndex.

See also
--------
set_names : able to set new names partially and by level

################################################################################
################################## Validation ##################################
################################################################################

Docstring for "pandas.Index.rename" correct. :)
```

```python
################################################################################
###################### Docstring (pandas.Index.set_names) ######################
################################################################################

Set new names on index. Defaults to returning new index.

Parameters
----------
names : str or sequence
    Name(s) to set.
level : int, level name, or sequence of int/level names (default None)
    If the index is a MultiIndex (hierarchical), level(s) to set (None
    for all levels).  Otherwise level must be None.
inplace : bool
    If True, mutates in place.

Returns
-------
new index (of same type and class...etc) [if inplace, returns None]

Examples
--------
>>> pd.Index([1, 2, 3, 4]).set_names('foo')
Int64Index([1, 2, 3, 4], dtype='int64', name='foo')
>>> pd.Index([1, 2, 3, 4]).set_names(['foo'])
Int64Index([1, 2, 3, 4], dtype='int64', name='foo')

>>> idx = pd.MultiIndex.from_tuples([(1, u'one'), (1, u'two'),
...                                  (2, u'one'), (2, u'two')],
...                                  names=['foo', 'bar'])
>>> idx.set_names(['baz', 'quz'])
MultiIndex(levels=[[1, 2], ['one', 'two']],
           labels=[[0, 0, 1, 1], [0, 1, 0, 1]],
           names=['baz', 'quz'])
>>> idx.set_names('baz', level=0)
MultiIndex(levels=[[1, 2], ['one', 'two']],
           labels=[[0, 0, 1, 1], [0, 1, 0, 1]],
           names=['baz', 'bar'])

See also
--------
rename : able to set new names without level

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	No extended summary found
```

jorisvandenbossche reviewed (approved):
Thanks a lot for the updates!

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1283,12 +1283,12 @@ def set_names(self, names, level=None, inplace=False):
         names : str or sequence
-            name(s) to set
+            Name(s) to set.
         level : int, level name, or sequence of int/level names (default None)
```
We try to have the types more than description in this line. Something like the next would be more standard:
`level: int, str, list of int or str, optional`

In the description you can expand on explaining that it is the level index or name.

https://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html#section-3-parameters

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1283,12 +1283,12 @@ def set_names(self, names, level=None, inplace=False):
+            If True, mutates in place.
 
         Returns
         -------
```
Do you mind changing the Return to a type too. Instead of "new index...", the type `Index` in the first line, and the description in the next.

https://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html#section-4-returns-or-yields

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1300,17 +1300,22 @@ def set_names(self, names, level=None, inplace=False):
+
+        See also
+        --------
+        rename : able to set new names without level
```
See also should have a capital `A` (`See Also`) and should be placed before the examples. Also, the description should start with upper case and finish with a period.

Is `rename` the method of `Index`? For sphinx to create a link to it, it should be `Index.rename`. Just rename would be understood as `pandas.rename` which does not exist.

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1341,16 +1346,39 @@ def rename(self, name, inplace=False):
+
+        See also
+        --------
+        set_names : able to set new names partially and by level
```
Same as before.

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1300,17 +1300,22 @@ def set_names(self, names, level=None, inplace=False):
         Int64Index([1, 2, 3, 4], dtype='int64', name='foo')
         >>> pd.Index([1, 2, 3, 4]).set_names(['foo'])
         Int64Index([1, 2, 3, 4], dtype='int64', name='foo')
```
These 2 are duplicated, aren't they?

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1300,17 +1300,22 @@ def set_names(self, names, level=None, inplace=False):
-                                            (2, u'one'), (2, u'two')],
-                                            names=['foo', 'bar'])
+        ...                                  (2, u'one'), (2, u'two')],
+        ...                                  names=['foo', 'bar'])
```
I think `names` is overindented by one space.

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1300,17 +1300,22 @@ def set_names(self, names, level=None, inplace=False):
         >>> idx = pd.MultiIndex.from_tuples([(1, u'one'), (1, u'two'),
-                                            (2, u'one'), (2, u'two')],
-                                            names=['foo', 'bar'])
+        ...                                  (2, u'one'), (2, u'two')],
```
I think we can get rid of the unicode `u` at the beginning of the strings.

datapythonista reviewed (changes requested):
Thanks for the PR, added some minor comments.

Moons08 commented:
I fixed detailed you said. and I changed one error msg to a proper one. 
And I got one question. 

now `Index.rename` get `level` parameter just `like set_names`. 
```python
>>> idx = pd.MultiIndex.from_tuples([(1, 'one'), (1, 'two'),
...                                  (2, 'one'), (2, 'two')],
...                                  names=['foo', 'bar'])
>>> idx.rename(1, level=0)
MultiIndex(levels=[[1, 2], ['one', 'two']],
           labels=[[0, 0, 1, 1], [0, 1, 0, 1]],
           names=[1, 'bar'])
```
should we prevent rename from getting level parameter?

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1278,39 +1278,46 @@ def _set_names(self, values, level=None):
-            if True, mutates in place
+        names : int, str, list-like, optional
+            Name(s) to set.
+        level : int, str, list-like, optional, default None
```
No need for the `default None` if we have `optional`

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1278,39 +1278,46 @@ def _set_names(self, values, level=None):
+        names : int, str, list-like, optional
+            Name(s) to set.
+        level : int, str, list-like, optional, default None
+            If the index is a MultiIndex(hierarchical), level(s) to set (None
```
I don't find the `hierarchical` necessary.

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1278,39 +1278,46 @@ def _set_names(self, values, level=None):
         Returns
         -------
-        new index (of same type and class...etc) [if inplace, returns None]
+        renamed : Index, MultiIndex or None
```
Just the type is enough, not need for the `renamed`

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1278,39 +1278,46 @@ def _set_names(self, values, level=None):
+
+        See also
+        --------
+        Index.rename : Able to set new names without level
```
Can you add a period at the end?

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1278,39 +1278,46 @@ def _set_names(self, values, level=None):
+        renamed : Index, MultiIndex or None
+            The same type as the caller or None if inplace is True.
+
+        See also
```
Capital `A` (`See Also`)

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1278,39 +1278,46 @@ def _set_names(self, values, level=None):
+
+        >>> idx = pd.MultiIndex.from_tuples([(1, 'one'), (1, 'two'),
+        ...                                  (2, 'one'), (2, 'two')],
+        ...                                  names=['foo', 'bar'])
```
I think in this case would help having a more real world example. At least for me is not so immediate to understand what we are dong, as replacing `foo` by `baz` is arbitrary. Having something like `['year', 'company']` changed to `['period', 'institution']` or something like this, IMO would make the example easier to understand.

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1278,39 +1278,46 @@ def _set_names(self, values, level=None):
         Examples
         --------
         >>> pd.Index([1, 2, 3, 4]).set_names('foo')
         Int64Index([1, 2, 3, 4], dtype='int64', name='foo')
```
I'd assign the index to a variable `idx`, display it, and then display it after setting the name. This example gives the impression that this is the right way to create an `Index` with a name.

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1319,7 +1326,8 @@ def set_names(self, names, level=None, inplace=False):
         if level is not None and not is_list_like(level) and is_list_like(
                 names):
-            raise TypeError("Names must be a string")
+            msg = "Length of names must match number of levels in MultiIndex."
```
I think the previous error message was more appropriate. May be better to say something like `Names must be a string when one level is provided`. But the new error message gives the impression that names should be a list, which is the not right.

Moons08 replied:
good. I will go for it

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1339,18 +1347,44 @@ def set_names(self, names, level=None, inplace=False):
+
+        See also
+        --------
+        Index.set_names : Able to set new names partially and by level
```
Same as before for `renamed`, `See Also` and the period.

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1339,18 +1347,44 @@ def set_names(self, names, level=None, inplace=False):
+                   names=['bar', None])
+        >>> idx.rename(['bar'])
+        Traceback (most recent call last):
+        ValueError: Length of names must match number of levels in MultiIndex.
```
Same as before regarding the examples.

datapythonista reviewed (changes requested):
Great work. Added some comments that IMO would make the docstring better.

jorisvandenbossche commented on pandas/core/indexes/base.py:
```diff
@@ -1278,39 +1278,46 @@ def _set_names(self, values, level=None):
-            for all levels).  Otherwise level must be None
-        inplace : bool
-            if True, mutates in place
+        names : int, str, list-like, optional
```
I don't think we should try to list the possible types here (`int, str`), as in principle an index name can be anything hashable, so not only ints or strings but also timestamps, tuples, ... 
I seem to remember that we had such a discussion during the documentation sprint recently, and we ended up using the generic "label" (and then it can be "label, or list of labels")

BTW, this one is not optional I think?

Moons08 replied:
right, it was ended with the generic "label". but @datapythonista said that we should follow this. 
https://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html#section-3-parameters

So I was not sure that using "label" in there is right. which one do you think better?
and yes, it is not optional. it because of my bad misunderstanding. I will delete it.

jorisvandenbossche replied:
Yeah, sorry, we should update that text

datapythonista replied:
Yes, sorry, I meant `object` instead of `str`, this is what we've been using for what I know. We tried to keep them as Python types, but I'm happy using `label`, it may be clearer. But if we use that, I'd prefer `list of label` to `list of labels`. This is consistent with `list of str` or `list of int`, and it will make it easier to parse and detect types automatically, if at some point that is useful.

datapythonista replied:
In this case I think `names` expects `label or list of label` if we go for the `label` option. Or `object or list of object` if we prefer the Python type. Am I right?

Moons08 replied:
I think `label or list of label` looks better

Moons08 commented:
I am really sorry for this... I am gonna fix it

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1278,39 +1278,49 @@ def _set_names(self, values, level=None):
-            if True, mutates in place
+        names : label or list of labels
+            Name(s) to set.
+        level : int, str or list of labels, optional
```
`label` instead of `str`

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1278,39 +1278,49 @@ def _set_names(self, values, level=None):
+
+        >>> idx = pd.MultiIndex.from_tuples([(1, 'one'), (1, 'two'),
+        ...                                  (2, 'one'), (2, 'two')],
+        ...                                  names=['year', 'company'])
```
I'd display `idx` after creating it. That will make it easier to compare with the result of `.set_names()`.

If we have `year` and `company`, I'd have something like `2018`, `Acme` in the values. I know I suggested it myself, but as I don't think we want to list company names, may be you can use something else?

Moons08 replied:
right, maybe we can use programming languages such as python, java, etc.

datapythonista replied:
I'd avoid any data that can lead to controversy, even if it's small. I use animals most of the times. :)

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -1319,7 +1329,8 @@ def set_names(self, names, level=None, inplace=False):
         if level is not None and not is_list_like(level) and is_list_like(
                 names):
-            raise TypeError("Names must be a string")
+            msg = "Names must be a string when one level is provided."
```
"when a single level" sounds clearer to me.

datapythonista reviewed (commented):
Looks great, thanks for all the changes. Just couple of small things.

datapythonista committed:
Minor changed to naming

datapythonista commented:
Thanks @Moons08, very good work, very clear docstring.
